<a href="https://colab.research.google.com/github/alicienty/HW2/blob/main/%D0%9B%D1%83%D0%BA%D1%8C%D1%8F%D0%BD%D1%87%D0%B8%D0%BA%D0%BE%D0%B2%D0%B0_trees_hw.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание: cравнение древовидных алгоритмов на NLP-данных

**Задача:** oбучить разные древовидные модели на текстах новостей, замерить test accuracy и время обучения, заполнить таблицы.

## 1. Загрузка и подготовка данных (код дан)

In [1]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import time
import pandas as pd

# Загружаем 3 категории
categories = ['rec.sport.baseball', 'sci.space', 'comp.graphics']
newsgroups = fetch_20newsgroups(subset='all', categories=categories, shuffle=True, random_state=42)

# TF-IDF векторизация
vectorizer = TfidfVectorizer(max_features=5000, ngram_range=(1, 2), stop_words='english')
X = vectorizer.fit_transform(newsgroups.data)
y = newsgroups.target

# Разделение на train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Размер обучающей выборки: {X_train.shape}")
print(f"Размер тестовой выборки: {X_test.shape}")

Размер обучающей выборки: (2363, 5000)
Размер тестовой выборки: (591, 5000)


## 2. Образец: Дерево решений (Decision Tree)

**Гиперпараметр:** `max_depth`

In [2]:
from sklearn.tree import DecisionTreeClassifier

results_tree = []

# Образец: мы делаем перебор значений max_depth = [3, 5, 10, 20, None])
# Для каждого: замер времени, обучение, accuracy на тесте, сохранение в results_tree

for depth in [3, 5, 10, 20, None]:
    start = time.time()
    clf = DecisionTreeClassifier(max_depth=depth, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_tree.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_tree = pd.DataFrame(results_tree)
print("\nТаблица 1. Дерево решений")
print(df_tree.to_markdown(index=False))

depth=3: acc=0.6244, time=0.17s
depth=5: acc=0.6971, time=0.27s
depth=10: acc=0.7716, time=0.87s
depth=20: acc=0.8409, time=0.83s
depth=None: acc=0.8697, time=0.93s

Таблица 1. Дерево решений
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           3 |          0.6244 |       0.17 |
|           5 |          0.6971 |       0.27 |
|          10 |          0.7716 |       0.87 |
|          20 |          0.8409 |       0.83 |
|         nan |          0.8697 |       0.93 |


## 3. Заполните таблицы для остальных алгоритмов

По аналогии с образцом обучите следующие модели и запишите результаты.

### 3.1. Случайный лес (Random Forest)

**Гиперпараметр:** `max_depth` (фиксируем `n_estimators=100`)

| max_depth | test_accuracy | time_sec |
|-----------|---------------|----------|
| 5         |               |          |
| 10        |               |          |
| 20        |               |          |
| None      |               |          |

In [15]:
from sklearn.ensemble import RandomForestClassifier

results_rf = []

# Ваш код здесь (перебор max_depth)
for depth in [5, 10, 20, None]:
    start = time.time()
    clf = RandomForestClassifier(max_depth=depth, random_state=42, n_estimators=100)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_rf.append({'max_depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_rf = pd.DataFrame(results_rf)
print("\nТаблица 2. Случайный лес")
print(df_rf.to_markdown(index=False))

depth=5: acc=0.9002, time=1.07s
depth=10: acc=0.9154, time=1.73s
depth=20: acc=0.9357, time=2.79s
depth=None: acc=0.9526, time=4.47s

Таблица 2. Случайный лес
|   max_depth |   test_accuracy |   time_sec |
|------------:|----------------:|-----------:|
|           5 |          0.9002 |       1.07 |
|          10 |          0.9154 |       1.73 |
|          20 |          0.9357 |       2.79 |
|         nan |          0.9526 |       4.47 |


### 3.2. Градиентный бустинг (Gradient Boosting)

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100, max_depth=3`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.01          |               |          |
| 0.05          |               |          |
| 0.1           |               |          |
| 0.5           |               |          |

In [28]:
from sklearn.ensemble import GradientBoostingClassifier

results_gb = []

# Ваш код здесь (перебор learning_rate)
for learning_rate in [0.01, 0.05, 0.1, 0.5]:
    start = time.time()
    clf = GradientBoostingClassifier(n_estimators=100, learning_rate=learning_rate, max_depth=3, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_gb.append({'learning_rate': learning_rate, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"learning_rate={learning_rate}: acc={acc:.4f}, time={fit_time:.2f}s")

df_gb = pd.DataFrame(results_gb)
print("\nТаблица 3. Градиентный бустинг")
print(df_gb.to_markdown(index=False))

learning_rate=0.01: acc=0.8782, time=36.21s
learning_rate=0.05: acc=0.9475, time=35.71s
learning_rate=0.1: acc=0.9577, time=35.41s
learning_rate=0.5: acc=0.9662, time=34.66s

Таблица 3. Градиентный бустинг
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|            0.01 |          0.8782 |      36.21 |
|            0.05 |          0.9475 |      35.71 |
|            0.1  |          0.9577 |      35.41 |
|            0.5  |          0.9662 |      34.66 |


### 3.3. XGBoost

**Гиперпараметр:** `gamma` (фиксируем `n_estimators=100, learning_rate=0.1`)

| gamma | test_accuracy | time_sec |
|-------|---------------|----------|
| 0     |               |          |
| 0.1   |               |          |
| 0.5   |               |          |
| 1.0   |               |          |

In [18]:
from xgboost import XGBClassifier

results_xgb = []

# Ваш код здесь (перебор gamma, use_label_encoder=False, eval_metric='logloss')
for gamma in [0, 0.1, 0.5, 1.0]:
    start = time.time()
    clf = XGBClassifier(gamma=gamma, n_estimators=100, learning_rate=0.1, eval_metric='logloss', random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_xgb.append({'gamma': gamma, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"gamma={gamma}: acc={acc:.4f}, time={fit_time:.2f}s")

df_xgb = pd.DataFrame(results_xgb)
print("\nТаблица 4. XGBoost")
print(df_xgb.to_markdown(index=False))

gamma=0: acc=0.9459, time=57.60s
gamma=0.1: acc=0.9442, time=42.76s
gamma=0.5: acc=0.9492, time=81.75s
gamma=1.0: acc=0.9492, time=50.07s

Таблица 4. XGBoost
|   gamma |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|     0   |          0.9459 |      57.6  |
|     0.1 |          0.9442 |      42.76 |
|     0.5 |          0.9492 |      81.75 |
|     1   |          0.9492 |      50.07 |


### 3.4. AdaBoost

**Гиперпараметр:** `learning_rate` (фиксируем `n_estimators=100`)

| learning_rate | test_accuracy | time_sec |
|---------------|---------------|----------|
| 0.5           |               |          |
| 1.0           |               |          |
| 1.5           |               |          |
| 2.0           |               |          |

In [14]:
from sklearn.ensemble import AdaBoostClassifier

results_ada = []

# Ваш код здесь (перебор learning_rate)
for learning_rate in [0.5, 1.0, 1.5, 2.0]:
    start = time.time()
    clf = AdaBoostClassifier(n_estimators=100, learning_rate=learning_rate, random_state=42)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_ada.append({'learning_rate': learning_rate, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"learning_rate={learning_rate}: acc={acc:.4f}, time={fit_time:.2f}s")

df_ada = pd.DataFrame(results_ada)
print("\nТаблица 5. AdaBoost")
print(df_ada.to_markdown(index=False))

learning_rate=0.5: acc=0.8629, time=13.15s
learning_rate=1.0: acc=0.9205, time=7.29s
learning_rate=1.5: acc=0.9255, time=5.31s
learning_rate=2.0: acc=0.8731, time=5.39s

Таблица 5. AdaBoost
|   learning_rate |   test_accuracy |   time_sec |
|----------------:|----------------:|-----------:|
|             0.5 |          0.8629 |      13.15 |
|             1   |          0.9205 |       7.29 |
|             1.5 |          0.9255 |       5.31 |
|             2   |          0.8731 |       5.39 |


### 3.5. LightGBM

**Гиперпараметр:** `num_leaves` (фиксируем `n_estimators=100, learning_rate=0.1`)

| num_leaves | test_accuracy | time_sec |
|------------|---------------|----------|
| 15         |               |          |
| 31         |               |          |
| 63         |               |          |
| 127        |               |          |

In [21]:
from lightgbm import LGBMClassifier

results_lgbm = []

# Ваш код здесь (перебор num_leaves, n_jobs=-1)
for num_leaves in [15, 31, 63, 127]:
    start = time.time()
    clf = LGBMClassifier(num_leaves=num_leaves, n_estimators=100, learning_rate=0.1, random_state=42, n_jobs=-1)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_lgbm.append({'num_leaves': num_leaves, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"num_leaves={num_leaves}: acc={acc:.4f}, time={fit_time:.2f}s")

df_lgbm = pd.DataFrame(results_lgbm)
print("\nТаблица 6. LightGBM")
print(df_lgbm.to_markdown(index=False))

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.405336 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=15: acc=0.9662, time=11.02s
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.084175 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=31: acc=0.9594, time=12.67s
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.114526 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


num_leaves=63: acc=0.9577, time=16.25s
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.085276 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 61063
[LightGBM] [Info] Number of data points in the train set: 2363, number of used features: 2963
[LightGBM] [Info] Start training from score -1.119999
[LightGBM] [Info] Start training from score -1.093126
[LightGBM] [Info] Start training from score -1.083076
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain,

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


### 3.6. CatBoost

**Гиперпараметр:** `depth` (фиксируем `iterations=100, learning_rate=0.1`)

| depth | test_accuracy | time_sec |
|-------|---------------|----------|
| 3     |               |          |
| 5     |               |          |
| 7     |               |          |
| 10    |               |          |

In [24]:
!pip install catboost

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.1/97.1 MB 9.0 MB/s eta 0:00:00


In [26]:
from catboost import CatBoostClassifier

results_cat = []

# Ваш код здесь (перебор depth, verbose=0)
for depth in [3, 5, 7, 10]:
    start = time.time()
    clf = CatBoostClassifier(depth=depth, iterations=100, learning_rate=0.1, random_state=42, verbose=0)
    clf.fit(X_train, y_train)
    fit_time = time.time() - start
    acc = accuracy_score(y_test, clf.predict(X_test))
    results_cat.append({'depth': depth, 'test_accuracy': round(acc, 4), 'time_sec': round(fit_time, 2)})
    print(f"depth={depth}: acc={acc:.4f}, time={fit_time:.2f}s")

df_cat = pd.DataFrame(results_cat)
print("\nТаблица 7. CatBoost")
print(df_cat.to_markdown(index=False))

depth=3: acc=0.9340, time=11.59s
depth=5: acc=0.9425, time=30.18s
depth=7: acc=0.9442, time=101.62s
depth=10: acc=0.9526, time=739.61s

Таблица 7. CatBoost
|   depth |   test_accuracy |   time_sec |
|--------:|----------------:|-----------:|
|       3 |          0.934  |      11.59 |
|       5 |          0.9425 |      30.18 |
|       7 |          0.9442 |     101.62 |
|      10 |          0.9526 |     739.61 |


## 4. Итоговая таблица лучших результатов

In [32]:
summary = pd.DataFrame([
    ['Decision Tree'] + df_tree.loc[df_tree.test_accuracy.idxmax()].tolist(),
    ['Random Forest'] + df_rf.loc[df_rf.test_accuracy.idxmax()].tolist(),
    ['Gradient Boosting'] + df_gb.loc[df_gb.test_accuracy.idxmax()].tolist(),
    ['XGBoost'] + df_xgb.loc[df_xgb.test_accuracy.idxmax()].tolist(),
    ['AdaBoost'] + df_ada.loc[df_ada.test_accuracy.idxmax()].tolist(),
    ['LightGBM'] + df_lgbm.loc[df_lgbm.test_accuracy.idxmax()].tolist(),
    ['CatBoost'] + df_cat.loc[df_cat.test_accuracy.idxmax()].tolist(),
], columns=['Algorithm', 'Best params', 'Best test acc', 'Time (sec)'])

print(summary.to_markdown(index=False))

| Algorithm         |   Best params |   Best test acc |   Time (sec) |
|:------------------|--------------:|----------------:|-------------:|
| Decision Tree     |         nan   |          0.8697 |         0.93 |
| Random Forest     |         nan   |          0.9526 |         4.47 |
| Gradient Boosting |           0.5 |          0.9662 |        34.66 |
| XGBoost           |           0.5 |          0.9492 |        81.75 |
| AdaBoost          |           1.5 |          0.9255 |         5.31 |
| LightGBM          |          15   |          0.9662 |        11.02 |
| CatBoost          |          10   |          0.9526 |       739.61 |


## 5. Вопросы (ответить текстом в следующей ячейке)

1. Какой алгоритм показал максимальную точность? Какие параметры к этому привели?

2. Какой алгоритм быстрее всего обучался? Во сколько раз он быстрее самого медленного?

3. У каких алгоритмов наблюдалось переобучение? При каких параметрах?

4. Какой алгоритм вы выбрали бы для продакшн-системы с миллионом текстов? Почему?

# Напишите свои ответы здесь

**Ответ 1:**

Максимальную точность показали Gradient Boosting и LightGBM (0.9662). Для градиентного бустинга лучший результат был достигнут при learning_rate=0.5, LightGBM — при num_leaves=15

**Ответ 2:**
Быстрее всего обучался Decision Tree (0.93 сек.). Он обучался быстрее CatBoost (739.61 сек) примерно в 795 раз. (пока обучался catboost я несколько раз успела подумать, что все сломалось)

**Ответ 3:**
Например, переобучение наблюдалось у LightGBM при повышении learning_rate до 31 (и выше), поскольку параметры росли, а качество работы снижалось.

Кроме того, у алгоритма CatBoost увеличение параметра depth увеличило время выполнения, но не дало значительного результата.

Также можно заметить, что алгоритм AdaBoost при learning_rate = 2.0 стал выдавать результаты хуже, что тоже говорит о переобучении.

**Ответ 4:**

Для такой системы мы должны выбрать модель с наилучими показателями качества (поскольку это бизнес и от этого будет зависеть прибыль компании), однако такую, которая будет работать сравнительно быстро. Я бы выбирала между Random Forest и LightGBM. Несмотря на то, что результаты у обеих моделей довольно высокие (0.95 и 0.97), LightGBM превосходит RF на 2 десятых. Тем не менее, RF сильно выигрывает LightGBM во времени (4.47 и 11.02). Выбирая RF, не слишком жертвуя качеством, мы выигрываем почти в 4 раза больше времени, что особенно важно при обработке миллионов текстов, когда важна каждая секунда (а то и меньше). Поэтому я бы остановилась на **RandomForest**, хотя LightGBM поддерживает многопоточность, что может ускорить время работы, требует дополнительной проверки и может считаться достойной альтернативой

## Критерии оценки

| Что оценивается | Баллы |
|----------------|-------|
| Заполнены все 7 таблиц (правильно собраны accuracy и время) | 3 |
| Заполнена итоговая таблица лучших результатов | 1 |
| Ответы на 4 вопроса (по 0.5 балла) | 2 |
| Код воспроизводим (random_state=42, порядок ячеек корректен) | 1 |
| **Качество кода** (отсутствие дублирования, осмысленные имена переменных, циклы вместо копипасты, использование .to_markdown()) | 3 |
| **Итого** | **10** |

### Детали по качеству кода (+3 балла):
- **+1** — использование единого шаблона для всех экспериментов (цикл + list.append + pd.DataFrame)
- **+1** — правильное форматирование вывода (округление времени и accuracy до 2-4 знаков)
- **+1** — читаемые имена переменных, комментарии, отсутствие `eval()`, `exec()` и магических чисел

**Штрафы:**
- -1 балл, если код не запускается без ошибок
- -1 балл, если отсутствует `random_state=42` в любой из моделей
- -1 балл, если таблицы выведены криво (не через `to_markdown` или нечитаемый `print`)